In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.options.display.max_columns = None
pd.options.display.max_rows = None

## I. Explore the data  
  
A. Study variable attributes 
 1. Identify variable name and survey item(s) it measures (Codebook available here: https://www.worldvaluessurvey.org/WVSDocumentationWV6.jsp (accessed on 5/25/2021))  
 2. % missing for each variable
 3. Quick descriptives (check for range of values, distribution shape, skew/outliers, potential errors, etc.)  
 4. Identify target variable (and drop duplicates) - don't forget to do feature engineering on the target variable; break out into varying levels of happiness  
  
B. Visualize the data (based on descriptives)
 1. Explore correlations between attributes
 2. Identify transformations that might be needed
 3. Identify extra data that may be useful (gini coefficient, GDP, etc.)
 4. Summarize findings

In [3]:
wvs_w6 = pd.read_csv('../data/Evaluating_Happiness/wvs_w6.csv', low_memory=False)

In [4]:
wvs_w6.shape

(89565, 442)

In [5]:
wvs_w6.head()

V1  V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA  V3   V4   V5   V6   V7   V8  \
0   6  12   12  615         ALG             DZA   1  1.0  1.0  1.0  NaN  1.0   
1   6  12   12  615         ALG             DZA   2  1.0  2.0  3.0  4.0  2.0   
2   6  12   12  615         ALG             DZA   3  1.0  3.0  2.0  4.0  2.0   
3   6  12   12  615         ALG             DZA   4  1.0  1.0  3.0  4.0  3.0   
4   6  12   12  615         ALG             DZA   5  1.0  1.0  1.0  2.0  1.0   

    V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  \
0  1.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  8.0   
1  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  5.0   
2  1.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  4.0   
3  1.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0  1.0  2.0  8.0   
4  1.0  1.0  3.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  8.0   

   V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  \
0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
3  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
4  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  1.0   

   V39  V40  V41  V42  V43  V44  V44_ES  V45  V46  V47  V48  V49  V50  V51  \
0  2.0  1.0  2.0  2.0  1.0  1.0     NaN  1.0  NaN  1.0  1.0  1.0  1.0  1.0   
1  2.0  2.0  2.0  2.0  2.0  2.0     NaN  1.0  NaN  1.0  1.0  2.0  2.0  1.0   
2  2.0  2.0  1.0  1.0  1.0  2.0     NaN  1.0  NaN  2.0  3.0  1.0  1.0  1.0   
3  2.0  2.0  2.0  2.0  1.0  2.0     NaN  2.0  NaN  1.0  1.0  1.0  2.0  2.0   
4  1.0  2.0  2.0  1.0  1.0  2.0     NaN  1.0  NaN  1.0  1.0  1.0  2.0  1.0   

   V52  V53  V54  V55  V56  V56_NZ  V57  V58   V59  V60  V61  V62  V63  V64  \
0  1.0  1.0  1.0  7.0  4.0     NaN  6.0  0.0  10.0  2.0  3.0  1.0  3.0  3.0   
1  2.0  3.0  2.0  6.0  8.0     NaN  6.0  0.0  10.0  2.0  1.0  2.0  3.0  4.0   
2  1.0  1.0  1.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  4.0  1.0  2.0  1.0   
3  2.0  2.0  2.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  1.0  3.0  1.0  4.0   
4  3.0  2.0  2.0  6.0  6.0     NaN  1.0  3.0   4.0  2.0  1.0  2.0  3.0  4.0   

   V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V74B  V75  V76  V77  V78  \
0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   2.0  2.0  1.0  2.0  3.0   
1  3.0  1.0  1.0  1.0  2.0  2.0  3.0  2.0  1.0  1.0   1.0  3.0  2.0  2.0  2.0   
2  4.0  1.0  2.0  2.0  2.0  1.0  2.0  1.0  1.0  4.0   3.0  1.0  2.0  2.0  2.0   
3  3.0  1.0  1.0  1.0  1.0  1.0  2.0  3.0  2.0  3.0   3.0  1.0  1.0  2.0  2.0   
4  2.0  1.0  2.0  1.0  1.0  2.0  5.0  1.0  2.0  3.0   1.0  4.0  3.0  2.0  2.0   

   V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  \
0  3.0  2.0  1.0  2.0  2.0  4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1  2.0  2.0  1.0  1.0  2.0  3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
2  2.0  2.0  1.0  2.0  2.0  2.0  NaN  NaN  3.0  3.0  3.0  NaN  NaN  NaN  NaN   
3  3.0  1.0  1.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
4  3.0  1.0  2.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  \
0  NaN  5.0  8.0  7.0  6.0  8.0   7.0   5.0   2.0   2.0   2.0   2.0   2.0   
1  NaN  5.0  7.0  5.0  5.0  4.0   4.0   6.0   1.0   3.0   3.0   3.0   3.0   
2  NaN  6.0  7.0  7.0  7.0  5.0   7.0   5.0   1.0   3.0   3.0   4.0   4.0   
3  NaN  6.0  9.0  5.0  6.0  4.0   6.0   8.0   1.0   3.0   3.0   2.0   2.0   
4  NaN  5.0  8.0  4.0  7.0  4.0   6.0   6.0   2.0   2.0   3.0   4.0   2.0   

   V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  \
0   2.0   1.0   3.0   3.0   2.0   3.0   2.0   2.0   4.0   3.0   3.0   2.0   
1   3.0   1.0   1.0   1.0   2.0

In [6]:
wvs_w6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89565 entries, 0 to 89564
Columns: 442 entries, V1 to WEIGHT4B
dtypes: float64(428), int64(12), object(2)
memory usage: 302.0+ MB


In [7]:
wvs_w6.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89565 entries, 0 to 89564
Data columns (total 442 columns):
 #   Column           Dtype  
---  ------           -----  
 0   V1               int64  
 1   V2               int64  
 2   V2A              int64  
 3   cow              int64  
 4   C_COW_ALPHA      object 
 5   B_COUNTRY_ALPHA  object 
 6   V3               int64  
 7   V4               float64
 8   V5               float64
 9   V6               float64
 10  V7               float64
 11  V8               float64
 12  V9               float64
 13  V10              float64
 14  V11              float64
 15  V12              float64
 16  V13              float64
 17  V14              float64
 18  V15              float64
 19  V16              float64
 20  V17              float64
 21  V18              float64
 22  V19              float64
 23  V20              float64
 24  V21              float64
 25  V22              float64
 26  V23              float64
 27  V24            

### Codebook notes:  
  
- Weights (See https://www.worldvaluessurvey.org/WVSContents.jsp for further details): 
    - S018 and S019 are weighting factors that transform N's to 1000 and 1500, respectively
    - these variables are useful for cross-country comparisons 
    - useful for EDA and descriptive analyses; should arguably be dropped for random forest algorithm, **right?**
    - **QUESTION:** are weights useful for PCA and logistic regression?
    - **QUESTION:** I see weights, but not specific population or sample size info - do I need this?
        - sample size shouldn't be too difficult to obtain based on max V3 values for each country
        - population data shouldn't be difficult to obtain based on N preserving weightings (V258) and this formula  
        
        $$Weight = S018/1000 * Population$$  

In [8]:
# sample sizes for each participating country (if needed)
wvs_w6.groupby('C_COW_ALPHA')['V3'].count()

C_COW_ALPHA
ALG    1200
ARG    1030
ARM    1100
AUL    1477
AZE    1002
BLR    1535
BRA    1486
CHL    1000
CHN    2300
COL    1512
CYP    1000
ECU    1202
EGY    1523
EST    1533
GHA    1552
GMY    2046
GRG    1202
HAI    1996
HKG    1000
IND    4078
IRQ    1200
JOR    1200
JPN    2443
KUW    1303
KYR    1500
KZK    1500
LEB    1200
LIB    2131
MAL    1300
MEX    2000
MOR    1200
NEW     841
NIG    1759
NTH    1902
PAK    1200
PER    1210
PHI    1200
POL     966
PSE    1000
QAT    1060
ROK    1200
ROM    1503
RUS    2500
RWA    1527
SAF    3531
SIN    1972
SLV    1069
SPN    1189
SWD    1206
TAW    1238
THI    1200
TRI     999
TUN    1205
TUR    1605
UKR    1500
URU    1000
USA    2232
UZB    1500
YEM    1000
ZIM    1500
Name: V3, dtype: int64

In [9]:
# how many columns with null values?
wvs_w6.isna().sum().count()

442

In [10]:
# another approach for finding the number of columns with null values
#sum(map(any, wvs_w6.isnull()))

In [11]:
# how many null values in each column?
wvs_w6.isna().sum()

V1                     0
V2                     0
V2A                    0
cow                    0
C_COW_ALPHA            0
B_COUNTRY_ALPHA        0
V3                     0
V4                   341
V5                   545
V6                  1035
V7                  1620
V8                  1519
V9                  1348
V10                  758
V11                  326
V12                   14
V13                    6
V14                    7
V15                   17
V16                    9
V17                   13
V18                   12
V19                   18
V20                   15
V21                   19
V22                    3
V23                  586
V24                 2388
V25                  840
V26                  923
V27                  993
V28                 3211
V29                 3129
V30                 1118
V31                 2223
V32                 1055
V33                 1214
V34                 1197
V35                 9696
V36                 3979


In [52]:
# how many missings by country?
grouped_missings = wvs_w6.groupby('C_COW_ALPHA')
grouped_missings = -grouped_missings.count().sub(grouped_missings.size(), 0)

V1  V2  V2A  cow  B_COUNTRY_ALPHA  V3  V4  V5   V6   V7   V8  \
C_COW_ALPHA                                                                 
ALG           0   0    0    0                0   0   5  11   18   63   31   
ARG           0   0    0    0                0   0   2   2    6   14   17   
ARM           0   0    0    0                0   0   7   5   26    7   18   
AUL           0   0    0    0                0   0   8  51   56   72   69   
AZE           0   0    0    0                0   0   0   0    0    0    0   
BLR           0   0    0    0                0   0   6  13   23   20   14   
BRA           0   0    0    0                0   0   1   3    6   11    3   
CHL           0   0    0    0                0   0   0  11    8   12   25   
CHN           0   0    0    0                0   0  29  51  101  190  107   
COL           0   0    0    0                0   0   0   2    2    7    2   
CYP           0   0    0    0                0   0   1   1    1   12    3   
ECU           0   0    0    0                0   0   0   0    1    0    0   
EGY           0   0    0    0                0   0   0   0    0    0    0   
EST           0   0    0    0                0   0   1   4    8    7   17   
GHA           0   0    0    0                0   0   0   0    0    0    0   
GMY           0   0    0    0                0   0   6   0    6    2   98   
GRG           0   0    0    0                0   0   0   0    1    1    1   
HAI           0   0    0    0                0   0  13  18   13   18   11   
HKG           0   0    0    0                0   0   2   0    5    7    2   
IND           0   0    0    0                0   0  17  67   74  166   69   
IRQ           0   0    0    0                0   0   0   2    9   11    2   
JOR           0   0    0    0                0   0   0   0    0    5   19   
JPN           0   0    0    0                0   0  39  54   69  177  132   
KUW           0   0    0    0                0   0  12  19   45   78   37   
KYR           0   0    0    0                0   0   0   0    0    0    0   
KZK           0   0    0    0                0   0   0   0    0    0    0   
LEB           0   0    0    0                0   0  55  16   22   67   34   
LIB           0   0    0    0                0   0   4  32   53   83   49   
MAL           0   0    0    0                0   0   0   0    0    1    0   
MEX           0   0    0    0                0   0   0   1    1    1    0   
MOR           0   0    0    0                0   0   2   4   81  108    9   
NEW           0   0    0    0                0   0  13  24   32   40   49   
NIG           0   0    0    0                0   0   0   0    0    1    0   
NTH           0   0    0    0                0   0  42  13   19   28   73   
PAK           0   0    0    0                0   0   1   2   16    8   19   
PER           0   0    0    0                0   0   0  13   20   32   19   
PHI           0   0    0    0                0   0   0   1    3    7    1   
POL           0   0    0    0                0   0   6   3   13    6   25   
PSE           0   0    0    0                0   0   1   2   11   15   29   
QAT           0   0    0    0                0   0   0   0    1    4    0   
ROK           0   0    0    0                0   0   1   3    6   12    9   
ROM           0   0    0    0                0   0   3   2    9   19    9   
RUS           0   0    0    0                0   0  14  21   46   62  191   
RWA           0   0    0    0                0   0   0   0    0    0    0   
SAF           0   0    0    0                0   0   5  11   32   62  125   
SIN           0   0    0    0                0   0   0   0    0    0    1   
SLV           0   0    0    0                0   0   5   1    8    4   13   
SPN           0   0    0    0                0   0   1  14   10    9   14   
SWD           0   0    0    0                0   0   3   2   14    9   67   
TAW           0   0    0    0                0   0   7  15   21   28   18   
THI      

In [12]:
# calculate proportion missing
pct_missing = pd.DataFrame(wvs_w6.isna().sum())
pct_missing = pct_missing.reset_index().rename(columns = {'index':'variable', 0:'NA_count'})
pct_missing

variable  NA_count
0                 V1         0
1                 V2         0
2                V2A         0
3                cow         0
4        C_COW_ALPHA         0
5    B_COUNTRY_ALPHA         0
6                 V3         0
7                 V4       341
8                 V5       545
9                 V6      1035
10                V7      1620
11                V8      1519
12                V9      1348
13               V10       758
14               V11       326
15               V12        14
16               V13         6
17               V14         7
18               V15        17
19               V16         9
20               V17        13
21               V18        12
22               V19        18
23               V20        15
24               V21        19
25               V22         3
26               V23       586
27               V24      2388
28               V25       840
29               V26       923
30               V27       993
31               V28      3211
32               V29      3129
33               V30      1118
34               V31      2223
35               V32      1055
36               V33      1214
37               V34      1197
38               V35      9696
39               V36      3979
40               V37      1541
41               V38      3982
42               V39      1546
43               V40      5281
44               V41      2841
45               V42      3974
46               V43      2841
47               V44        20
48            V44_ES     88376
49               V45      1616
50               V46      7607
51               V47      3456
52               V48      2966
53               V49      3193
54               V50      4048
55               V51      4098
56               V52      2605
57               V53      3912
58               V54      4656
59               V55      1448
60               V56      2726
61            V56_NZ     88782
62               V57       244
63               V58      1667
64               V59       648
65               V60      3281
66               V61      5617
67               V62      2937
68               V63      4835
69               V64      2812
70               V65      4526
71               V66      8905
72               V67      3274
73               V68      3466
74               V69      3784
75               V70      2977
76               V71      2595
77               V72      2091
78               V73      2486
79               V74      5602
80              V74B     49678
81               V75      2844
82               V76      3132
83               V77      2336
84               V78      2326
85               V79      2121
86               V80      1115
87               V81      6548
88               V82      1055
89               V83      1351
90               V84       997
91               V85     10624
92               V86     12961
93               V87     10790
94               V88     11317
95               V89     19961
96               V90     74866
97               V91     84618
98               V92     79341
99               V93     83127
100              V94     85377
101              V95     21366
102              V96      2885
103              V97      5298
104              V98      2253
105              V99      3154
106             V100      2258
107             V101      4602
108             V102      1455
109             V103      2063
110             V104      1955
111             V105      3130
112             V106      6956
113             V107      7189
114             V108      2927
115             V109      4397
116             V110      2783
117             V111      2019
118             V112     11170
119             V113      2078
120             V114      3338
121             V115      3178
122             V116      6353
123             V117      4497
124             V118      5655
125             V119      5372
126             V120      6468
127             V121      4023
128          

In [13]:
pct_missing['NA_pct'] = (pct_missing['NA_count'] / len(wvs_w6)) * 100
pct_missing['NA_pct'] = pct_missing['NA_pct'].round(decimals=2)
pct_missing

variable  NA_count  NA_pct
0                 V1         0    0.00
1                 V2         0    0.00
2                V2A         0    0.00
3                cow         0    0.00
4        C_COW_ALPHA         0    0.00
5    B_COUNTRY_ALPHA         0    0.00
6                 V3         0    0.00
7                 V4       341    0.38
8                 V5       545    0.61
9                 V6      1035    1.16
10                V7      1620    1.81
11                V8      1519    1.70
12                V9      1348    1.51
13               V10       758    0.85
14               V11       326    0.36
15               V12        14    0.02
16               V13         6    0.01
17               V14         7    0.01
18               V15        17    0.02
19               V16         9    0.01
20               V17        13    0.01
21               V18        12    0.01
22               V19        18    0.02
23               V20        15    0.02
24               V21        19    0.02
25               V22         3    0.00
26               V23       586    0.65
27               V24      2388    2.67
28               V25       840    0.94
29               V26       923    1.03
30               V27       993    1.11
31               V28      3211    3.59
32               V29      3129    3.49
33               V30      1118    1.25
34               V31      2223    2.48
35               V32      1055    1.18
36               V33      1214    1.36
37               V34      1197    1.34
38               V35      9696   10.83
39               V36      3979    4.44
40               V37      1541    1.72
41               V38      3982    4.45
42               V39      1546    1.73
43               V40      5281    5.90
44               V41      2841    3.17
45               V42      3974    4.44
46               V43      2841    3.17
47               V44        20    0.02
48            V44_ES     88376   98.67
49               V45      1616    1.80
50               V46      7607    8.49
51               V47      3456    3.86
52               V48      2966    3.31
53               V49      3193    3.57
54               V50      4048    4.52
55               V51      4098    4.58
56               V52      2605    2.91
57               V53      3912    4.37
58               V54      4656    5.20
59               V55      1448    1.62
60               V56      2726    3.04
61            V56_NZ     88782   99.13
62               V57       244    0.27
63               V58      1667    1.86
64               V59       648    0.72
65               V60      3281    3.66
66               V61      5617    6.27
67               V62      2937    3.28
68               V63      4835    5.40
69               V64      2812    3.14
70               V65      4526    5.05
71               V66      8905    9.94
72               V67      3274    3.66
73               V68      3466    3.87
74               V69      3784    4.22
75               V70      2977    3.32
76               V71      2595    2.90
77               V72      2091    2.33
78               V73      2486    2.78
79               V74      5602    6.25
80              V74B     49678   55.47
81               V75      2844    3.18
82               V76      3132    3.50
83               V77      2336    2.61
84               V78      2326    2.60
85               V79      2121    2.37
86               V80      1115    1.24
87               V81      6548    7.31
88               V82      1055    1.18
89               V83      1351    1.51
90               V84       997    1.11
91               V85     10624   11.86
92               V86     12961   14.47
93               V87     10790   12.05
94               V88     11317   12.64
95               V89     19961   22.29
96               V90     74866   83.59
97               V91     84618   94.48
98               V92     79341   88.58
99               V93     83127   92.81
100              V94     85377   95.32
101              V95     21366   2

In [14]:
# how many columns contain missing values?
print(np.count_nonzero(wvs_w6.isna().sum() != 0), 'columns contain missing values')

# how many columns do not?
print(np.count_nonzero(wvs_w6.isna().sum() == 0), 'columns DO NOT contain missing values')

414 columns contain missing values
28 columns DO NOT contain missing values


#### Observations so far:
This data set contains 28 columns without any missing information, and 414 columns (out of 442) that contain at least 1 missing value.

#### Next steps:
- figure out if these missing values matter for my analyses
- decide what to do about missing values

In [15]:
# list columns containing missing data
wvs_w6.columns[wvs_w6.isna().any()].tolist()

['V4',
 'V5',
 'V6',
 'V7',
 'V8',
 'V9',
 'V10',
 'V11',
 'V12',
 'V13',
 'V14',
 'V15',
 'V16',
 'V17',
 'V18',
 'V19',
 'V20',
 'V21',
 'V22',
 'V23',
 'V24',
 'V25',
 'V26',
 'V27',
 'V28',
 'V29',
 'V30',
 'V31',
 'V32',
 'V33',
 'V34',
 'V35',
 'V36',
 'V37',
 'V38',
 'V39',
 'V40',
 'V41',
 'V42',
 'V43',
 'V44',
 'V44_ES',
 'V45',
 'V46',
 'V47',
 'V48',
 'V49',
 'V50',
 'V51',
 'V52',
 'V53',
 'V54',
 'V55',
 'V56',
 'V56_NZ',
 'V57',
 'V58',
 'V59',
 'V60',
 'V61',
 'V62',
 'V63',
 'V64',
 'V65',
 'V66',
 'V67',
 'V68',
 'V69',
 'V70',
 'V71',
 'V72',
 'V73',
 'V74',
 'V74B',
 'V75',
 'V76',
 'V77',
 'V78',
 'V79',
 'V80',
 'V81',
 'V82',
 'V83',
 'V84',
 'V85',
 'V86',
 'V87',
 'V88',
 'V89',
 'V90',
 'V91',
 'V92',
 'V93',
 'V94',
 'V95',
 'V96',
 'V97',
 'V98',
 'V99',
 'V100',
 'V101',
 'V102',
 'V103',
 'V104',
 'V105',
 'V106',
 'V107',
 'V108',
 'V109',
 'V110',
 'V111',
 'V112',
 'V113',
 'V114',
 'V115',
 'V116',
 'V117',
 'V118',
 'V119',
 'V120',
 'V121',
 'V122',


In [16]:
# view variables containing missing values and % missing
pct_missing[pct_missing['NA_count'] != 0]

variable  NA_count  NA_pct
7               V4       341    0.38
8               V5       545    0.61
9               V6      1035    1.16
10              V7      1620    1.81
11              V8      1519    1.70
12              V9      1348    1.51
13             V10       758    0.85
14             V11       326    0.36
15             V12        14    0.02
16             V13         6    0.01
17             V14         7    0.01
18             V15        17    0.02
19             V16         9    0.01
20             V17        13    0.01
21             V18        12    0.01
22             V19        18    0.02
23             V20        15    0.02
24             V21        19    0.02
25             V22         3    0.00
26             V23       586    0.65
27             V24      2388    2.67
28             V25       840    0.94
29             V26       923    1.03
30             V27       993    1.11
31             V28      3211    3.59
32             V29      3129    3.49
33             V30      1118    1.25
34             V31      2223    2.48
35             V32      1055    1.18
36             V33      1214    1.36
37             V34      1197    1.34
38             V35      9696   10.83
39             V36      3979    4.44
40             V37      1541    1.72
41             V38      3982    4.45
42             V39      1546    1.73
43             V40      5281    5.90
44             V41      2841    3.17
45             V42      3974    4.44
46             V43      2841    3.17
47             V44        20    0.02
48          V44_ES     88376   98.67
49             V45      1616    1.80
50             V46      7607    8.49
51             V47      3456    3.86
52             V48      2966    3.31
53             V49      3193    3.57
54             V50      4048    4.52
55             V51      4098    4.58
56             V52      2605    2.91
57             V53      3912    4.37
58             V54      4656    5.20
59             V55      1448    1.62
60             V56      2726    3.04
61          V56_NZ     88782   99.13
62             V57       244    0.27
63             V58      1667    1.86
64             V59       648    0.72
65             V60      3281    3.66
66             V61      5617    6.27
67             V62      2937    3.28
68             V63      4835    5.40
69             V64      2812    3.14
70             V65      4526    5.05
71             V66      8905    9.94
72             V67      3274    3.66
73             V68      3466    3.87
74             V69      3784    4.22
75             V70      2977    3.32
76             V71      2595    2.90
77             V72      2091    2.33
78             V73      2486    2.78
79             V74      5602    6.25
80            V74B     49678   55.47
81             V75      2844    3.18
82             V76      3132    3.50
83             V77      2336    2.61
84             V78      2326    2.60
85             V79      2121    2.37
86             V80      1115    1.24
87             V81      6548    7.31
88             V82      1055    1.18
89             V83      1351    1.51
90             V84       997    1.11
91             V85     10624   11.86
92             V86     12961   14.47
93             V87     10790   12.05
94             V88     11317   12.64
95             V89     19961   22.29
96             V90     74866   83.59
97             V91     84618   94.48
98             V92     79341   88.58
99             V93     83127   92.81
100            V94     85377   95.32
101            V95     21366   23.86
102            V96      2885    3.22
103            V97      5298    5.92
104            V98      2253    2.52
105            V99      3154    3.52
106           V100      2258    2.52
107           V101      4602    5.14
108           V102      1455    1.62
109           V103      2063    2.30
110           V104      1955    2.18
111           V105      3130    3.49
112           V106      6956    7.77
113           V107      7189    8.03
114           

In [17]:
# create list of variables where pct missing is 50% or higher
half_missing = pct_missing[pct_missing['NA_pct'] >= 50.00]['variable'].tolist()
half_missing

['V44_ES',
 'V56_NZ',
 'V74B',
 'V90',
 'V91',
 'V92',
 'V93',
 'V94',
 'V125_00',
 'V125_01',
 'V125_02',
 'V125_03',
 'V125_04',
 'V125_05',
 'V125_06',
 'V125_07',
 'V125_08',
 'V125_09',
 'V125_10',
 'V125_11',
 'V125_12',
 'V125_13',
 'V125_14',
 'V125_15',
 'V160A',
 'V160B',
 'V160C',
 'V160D',
 'V160E',
 'V160F',
 'V160G',
 'V160H',
 'V160I',
 'V160J',
 'V207A',
 'V215_01',
 'V215_02',
 'V215_03',
 'V215_04',
 'V215_05',
 'V215_06',
 'V215_07',
 'V215_08',
 'V215_10',
 'V215_11',
 'V215_12',
 'V215_13',
 'V215_14',
 'V215_15',
 'V215_16',
 'V215_17',
 'V215_18',
 'V218_ESMA',
 'V217_ESMA',
 'V219_ESMA',
 'V220_ESMA',
 'V221_ESMA',
 'V222_ESMA',
 'V223_ESMA',
 'V224_ESMA',
 'V228_2',
 'V243_AU',
 'V244_AU',
 'V248_CS',
 'V253_CS',
 'V256B',
 'V256C',
 'V265',
 'MN_35A',
 'MN_163A',
 'MN_163B',
 'MN_163C',
 'MN_228L',
 'MN_228M',
 'MN_228N',
 'MN_228O',
 'MN_228P',
 'MN_228Q',
 'MN_228R',
 'MN_228S1',
 'MN_228S2',
 'MN_228S3',
 'MN_228S4',
 'MN_228S5',
 'MN_228S6',
 'MN_228S7',
 

### Options for handling missing data:  
Advice from https://heartbeat.fritz.ai/data-handling-scenarios-part-2-working-with-missing-values-in-a-dataset-34b758cfc9fa and https://analyticsindiamag.com/5-ways-handle-missing-values-machine-learning-datasets/  
  
**Drop rows and columns with missing data**  
1. pros: 
  - quick and easy approach 
2. cons: 
  - could mean losing important information about the data along with the missing values
  - when using your model in production, the model will not automatically know how to handle missing data
3. when this makes sense: 
  - rows containing NULL values (missing values) are around 5% (or less) of the total data
  - NULL values in columns are significantly more than the other values present; it wouldn’t make sense to keep these, as they hold little or no descriptive information about the data  
  
**Mean/Median (numerical) & Mode (categorical) imputation**  
1. pros: 
  - easy to do
  - can be integrated into production or for a future unknown dataset
2. cons: 
  - distorts the distribution of the dataset
  - distorts the variance and covariance of the dataset
  - for mode imputation, may lead to an over-representation of the most frequent label if the missing values are quite large
3. when this makes sense: 
  - mean imputation works best for normally distributed distributions
  - median is better for skewed distributions 
  - mode imputation for categorical data works best if the missing values are missing at random
  - best to use this method when the missing values are around 5% (or less) of the total data
  
**Systematic Random Sampling Imputation**  
1. pros: 
  - does not distort variance or distribution 
2. cons: 
  - when replacing missing values in the test set as well, the imputed values from the train set will need to be stored in memory
3. when this makes sense: 
  - can be applied to both numerical and categorical variables
  - used when the values are missing at random
  - when we want to be able to reproduce the same value every time the variable is used (by using a random state)
  
### Remaining Questions:
Is "stratified" mean/median imputation an option?

In [18]:
# first step in droping columns with 50% or more missing observations
half_missing

['V44_ES',
 'V56_NZ',
 'V74B',
 'V90',
 'V91',
 'V92',
 'V93',
 'V94',
 'V125_00',
 'V125_01',
 'V125_02',
 'V125_03',
 'V125_04',
 'V125_05',
 'V125_06',
 'V125_07',
 'V125_08',
 'V125_09',
 'V125_10',
 'V125_11',
 'V125_12',
 'V125_13',
 'V125_14',
 'V125_15',
 'V160A',
 'V160B',
 'V160C',
 'V160D',
 'V160E',
 'V160F',
 'V160G',
 'V160H',
 'V160I',
 'V160J',
 'V207A',
 'V215_01',
 'V215_02',
 'V215_03',
 'V215_04',
 'V215_05',
 'V215_06',
 'V215_07',
 'V215_08',
 'V215_10',
 'V215_11',
 'V215_12',
 'V215_13',
 'V215_14',
 'V215_15',
 'V215_16',
 'V215_17',
 'V215_18',
 'V218_ESMA',
 'V217_ESMA',
 'V219_ESMA',
 'V220_ESMA',
 'V221_ESMA',
 'V222_ESMA',
 'V223_ESMA',
 'V224_ESMA',
 'V228_2',
 'V243_AU',
 'V244_AU',
 'V248_CS',
 'V253_CS',
 'V256B',
 'V256C',
 'V265',
 'MN_35A',
 'MN_163A',
 'MN_163B',
 'MN_163C',
 'MN_228L',
 'MN_228M',
 'MN_228N',
 'MN_228O',
 'MN_228P',
 'MN_228Q',
 'MN_228R',
 'MN_228S1',
 'MN_228S2',
 'MN_228S3',
 'MN_228S4',
 'MN_228S5',
 'MN_228S6',
 'MN_228S7',
 

In [19]:
len(half_missing)

110

#### Re-insert variables that shouldn't be dropped
- V74B (Schwartz value scale item)
- V207A (euthanasia justifiable); curious to see if this could be imputed
- V215_01 - V215_18; use these to create a "sense of belonging to political/economic union" variable (V215_B)

In [20]:
# create a for loop to re-insert variables into the list that shouldn't be dropped (at this point)
# original list: high_missings_list
# new list
vars_to_keep = ['V56_NZ', 'V74B', 'V207A', 'V215_01', 'V215_02', 'V215_03', 'V215_04', 'V215_05', 
                'V215_06', 'V215_07', 'V215_08', 'V215_10', 'V215_11', 'V215_12', 'V215_13', 'V215_14', 
                'V215_15', 'V215_16', 'V215_17', 'V215_18']

vars_to_drop = []

for variable in half_missing:
    if variable not in vars_to_keep:
        
        vars_to_drop.append(variable)

print(vars_to_drop)

['V44_ES', 'V90', 'V91', 'V92', 'V93', 'V94', 'V160A', 'V160B', 'V160C', 'V160D', 'V160E', 'V160F', 'V160G', 'V160H', 'V160I', 'V160J', 'V218_ESMA', 'V217_ESMA', 'V219_ESMA', 'V220_ESMA', 'V221_ESMA', 'V222_ESMA', 'V223_ESMA', 'V224_ESMA', 'V228_2', 'V243_AU', 'V244_AU', 'V248_CS', 'V253_CS', 'V256B', 'V256C', 'V265', 'MN_35A', 'MN_163A', 'MN_163B', 'MN_163C', 'MN_228L', 'MN_228M', 'MN_228N', 'MN_228O', 'MN_228P', 'MN_228Q', 'MN_228R', 'MN_228S1', 'MN_228S2', 'MN_228S3', 'MN_228S4', 'MN_228S5', 'MN_228S6', 'MN_228S7', 'MN_228S8', 'MN_229A', 'MN_229B', 'MN_230A', 'MN_233A', 'MN_233B', 'MN_234A', 'MN_237A', 'MN_237B1', 'MN_237B2', 'MN_237B3', 'MN_237B4', 'MN_237B5', 'MN_237B6', 'MN_237B7', 'MN_237C1', 'MN_237C2', 'MN_237C3', 'MN_237C4', 'MN_237C5', 'MN_237C6', 'MN_249A1', 'MN_249A2', 'MN_249A3']


In [21]:
len(vars_to_drop)

74

In [22]:
# view original imported data set
wvs_w6.head()

V1  V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA  V3   V4   V5   V6   V7   V8  \
0   6  12   12  615         ALG             DZA   1  1.0  1.0  1.0  NaN  1.0   
1   6  12   12  615         ALG             DZA   2  1.0  2.0  3.0  4.0  2.0   
2   6  12   12  615         ALG             DZA   3  1.0  3.0  2.0  4.0  2.0   
3   6  12   12  615         ALG             DZA   4  1.0  1.0  3.0  4.0  3.0   
4   6  12   12  615         ALG             DZA   5  1.0  1.0  1.0  2.0  1.0   

    V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  \
0  1.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  8.0   
1  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  5.0   
2  1.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  4.0   
3  1.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0  1.0  2.0  8.0   
4  1.0  1.0  3.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  8.0   

   V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  \
0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
3  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
4  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  1.0   

   V39  V40  V41  V42  V43  V44  V44_ES  V45  V46  V47  V48  V49  V50  V51  \
0  2.0  1.0  2.0  2.0  1.0  1.0     NaN  1.0  NaN  1.0  1.0  1.0  1.0  1.0   
1  2.0  2.0  2.0  2.0  2.0  2.0     NaN  1.0  NaN  1.0  1.0  2.0  2.0  1.0   
2  2.0  2.0  1.0  1.0  1.0  2.0     NaN  1.0  NaN  2.0  3.0  1.0  1.0  1.0   
3  2.0  2.0  2.0  2.0  1.0  2.0     NaN  2.0  NaN  1.0  1.0  1.0  2.0  2.0   
4  1.0  2.0  2.0  1.0  1.0  2.0     NaN  1.0  NaN  1.0  1.0  1.0  2.0  1.0   

   V52  V53  V54  V55  V56  V56_NZ  V57  V58   V59  V60  V61  V62  V63  V64  \
0  1.0  1.0  1.0  7.0  4.0     NaN  6.0  0.0  10.0  2.0  3.0  1.0  3.0  3.0   
1  2.0  3.0  2.0  6.0  8.0     NaN  6.0  0.0  10.0  2.0  1.0  2.0  3.0  4.0   
2  1.0  1.0  1.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  4.0  1.0  2.0  1.0   
3  2.0  2.0  2.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  1.0  3.0  1.0  4.0   
4  3.0  2.0  2.0  6.0  6.0     NaN  1.0  3.0   4.0  2.0  1.0  2.0  3.0  4.0   

   V65  V66  V67  V68  V69  V70  V71  V72  V73  V74  V74B  V75  V76  V77  V78  \
0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   2.0  2.0  1.0  2.0  3.0   
1  3.0  1.0  1.0  1.0  2.0  2.0  3.0  2.0  1.0  1.0   1.0  3.0  2.0  2.0  2.0   
2  4.0  1.0  2.0  2.0  2.0  1.0  2.0  1.0  1.0  4.0   3.0  1.0  2.0  2.0  2.0   
3  3.0  1.0  1.0  1.0  1.0  1.0  2.0  3.0  2.0  3.0   3.0  1.0  1.0  2.0  2.0   
4  2.0  1.0  2.0  1.0  1.0  2.0  5.0  1.0  2.0  3.0   1.0  4.0  3.0  2.0  2.0   

   V79  V80  V81  V82  V83  V84  V85  V86  V87  V88  V89  V90  V91  V92  V93  \
0  3.0  2.0  1.0  2.0  2.0  4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1  2.0  2.0  1.0  1.0  2.0  3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
2  2.0  2.0  1.0  2.0  2.0  2.0  NaN  NaN  3.0  3.0  3.0  NaN  NaN  NaN  NaN   
3  3.0  1.0  1.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
4  3.0  1.0  2.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V94  V95  V96  V97  V98  V99  V100  V101  V102  V103  V104  V105  V106  \
0  NaN  5.0  8.0  7.0  6.0  8.0   7.0   5.0   2.0   2.0   2.0   2.0   2.0   
1  NaN  5.0  7.0  5.0  5.0  4.0   4.0   6.0   1.0   3.0   3.0   3.0   3.0   
2  NaN  6.0  7.0  7.0  7.0  5.0   7.0   5.0   1.0   3.0   3.0   4.0   4.0   
3  NaN  6.0  9.0  5.0  6.0  4.0   6.0   8.0   1.0   3.0   3.0   2.0   2.0   
4  NaN  5.0  8.0  4.0  7.0  4.0   6.0   6.0   2.0   2.0   3.0   4.0   2.0   

   V107  V108  V109  V110  V111  V112  V113  V114  V115  V116  V117  V118  \
0   2.0   1.0   3.0   3.0   2.0   3.0   2.0   2.0   4.0   3.0   3.0   2.0   
1   3.0   1.0   1.0   1.0   2.0

In [23]:
# drop variables in the vars_to_drop list
cleaning_w6 = wvs_w6.drop(columns=vars_to_drop)

In [24]:
# check out current freq of missing values
cleaning_w6.isna().sum()

V1                     0
V2                     0
V2A                    0
cow                    0
C_COW_ALPHA            0
B_COUNTRY_ALPHA        0
V3                     0
V4                   341
V5                   545
V6                  1035
V7                  1620
V8                  1519
V9                  1348
V10                  758
V11                  326
V12                   14
V13                    6
V14                    7
V15                   17
V16                    9
V17                   13
V18                   12
V19                   18
V20                   15
V21                   19
V22                    3
V23                  586
V24                 2388
V25                  840
V26                  923
V27                  993
V28                 3211
V29                 3129
V30                 1118
V31                 2223
V32                 1055
V33                 1214
V34                 1197
V35                 9696
V36                 3979


In [25]:
# peek at data to ensure that correct columns are being dropped so far
cleaning_w6.head()

V1  V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA  V3   V4   V5   V6   V7   V8  \
0   6  12   12  615         ALG             DZA   1  1.0  1.0  1.0  NaN  1.0   
1   6  12   12  615         ALG             DZA   2  1.0  2.0  3.0  4.0  2.0   
2   6  12   12  615         ALG             DZA   3  1.0  3.0  2.0  4.0  2.0   
3   6  12   12  615         ALG             DZA   4  1.0  1.0  3.0  4.0  3.0   
4   6  12   12  615         ALG             DZA   5  1.0  1.0  1.0  2.0  1.0   

    V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  \
0  1.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  8.0   
1  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  5.0   
2  1.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  4.0   
3  1.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0  1.0  2.0  8.0   
4  1.0  1.0  3.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  8.0   

   V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  \
0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
3  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
4  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  1.0   

   V39  V40  V41  V42  V43  V44  V45  V46  V47  V48  V49  V50  V51  V52  V53  \
0  2.0  1.0  2.0  2.0  1.0  1.0  1.0  NaN  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
1  2.0  2.0  2.0  2.0  2.0  2.0  1.0  NaN  1.0  1.0  2.0  2.0  1.0  2.0  3.0   
2  2.0  2.0  1.0  1.0  1.0  2.0  1.0  NaN  2.0  3.0  1.0  1.0  1.0  1.0  1.0   
3  2.0  2.0  2.0  2.0  1.0  2.0  2.0  NaN  1.0  1.0  1.0  2.0  2.0  2.0  2.0   
4  1.0  2.0  2.0  1.0  1.0  2.0  1.0  NaN  1.0  1.0  1.0  2.0  1.0  3.0  2.0   

   V54  V55  V56  V56_NZ  V57  V58   V59  V60  V61  V62  V63  V64  V65  V66  \
0  1.0  7.0  4.0     NaN  6.0  0.0  10.0  2.0  3.0  1.0  3.0  3.0  1.0  1.0   
1  2.0  6.0  8.0     NaN  6.0  0.0  10.0  2.0  1.0  2.0  3.0  4.0  3.0  1.0   
2  1.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  4.0  1.0  2.0  1.0  4.0  1.0   
3  2.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  1.0  3.0  1.0  4.0  3.0  1.0   
4  2.0  6.0  6.0     NaN  1.0  3.0   4.0  2.0  1.0  2.0  3.0  4.0  2.0  1.0   

   V67  V68  V69  V70  V71  V72  V73  V74  V74B  V75  V76  V77  V78  V79  V80  \
0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   2.0  2.0  1.0  2.0  3.0  3.0  2.0   
1  1.0  1.0  2.0  2.0  3.0  2.0  1.0  1.0   1.0  3.0  2.0  2.0  2.0  2.0  2.0   
2  2.0  2.0  2.0  1.0  2.0  1.0  1.0  4.0   3.0  1.0  2.0  2.0  2.0  2.0  2.0   
3  1.0  1.0  1.0  1.0  2.0  3.0  2.0  3.0   3.0  1.0  1.0  2.0  2.0  3.0  1.0   
4  2.0  1.0  1.0  2.0  5.0  1.0  2.0  3.0   1.0  4.0  3.0  2.0  2.0  3.0  1.0   

   V81  V82  V83  V84  V85  V86  V87  V88  V89  V95  V96  V97  V98  V99  V100  \
0  1.0  2.0  2.0  4.0  NaN  NaN  NaN  NaN  NaN  5.0  8.0  7.0  6.0  8.0   7.0   
1  1.0  1.0  2.0  3.0  NaN  NaN  NaN  NaN  NaN  5.0  7.0  5.0  5.0  4.0   4.0   
2  1.0  2.0  2.0  2.0  NaN  NaN  3.0  3.0  3.0  6.0  7.0  7.0  7.0  5.0   7.0   
3  1.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  6.0  9.0  5.0  6.0  4.0   6.0   
4  2.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  5.0  8.0  4.0  7.0  4.0   6.0   

   V101  V102  V103  V104  V105  V106  V107  V108  V109  V110  V111  V112  \
0   5.0   2.0   2.0   2.0   2.0   2.0   2.0   1.0   3.0   3.0   2.0   3.0   
1   6.0   1.0   3.0   3.0   3.0   3.0   3.0   1.0   1.0   1.0   2.0   2.0   
2   5.0   1.0   3.0   3.0   4.0   4.0   4.0   3.0   2.0   2.0   2.0   3.0   
3   8.0   1.0   3.0   3.0   2.0   2.0   3.0   2.0   3.0   4.0   2.0   3.0   
4   6.0   2.0   2.0   3.0   4.0   2.0   3.0   2.0   3.0   3.0   2.0   4.0   

   V113  V114  V115  V116  V117  V118  V119  V120  V121  V122  V123  V124  \
0   2.0   2.0   4.0   3.0   3.0   2.0   3.0   2.0   3.0   4.0   4.0   3.0   
1   3.0   1.0

In [26]:
# original shape for comparison
wvs_w6.shape

(89565, 442)

In [27]:
# current shape so far
cleaning_w6.shape

(89565, 368)

In [28]:
# Columns to drop
# drop a handful of item scales at the end of the data set since these come from items that appear earlier in the survey
# also dropping their associated weights
cleaning_w6 = cleaning_w6.drop(['Y001_1', 'Y001_2', 'Y001_3', 'Y001_4', 'Y001_5', 'Y001', 'Y002', 'Y003', 'sacsecval',
 'secvalwgt', 'resemaval', 'weightb', 'I_AUTHORITY', 'I_NATIONALISM', 'I_DEVOUT', 'defiance', 'I_RELIGIMP', 'I_RELIGBEL',
 'I_RELIGPRAC', 'disbelief', 'I_NORM1', 'I_NORM2', 'I_NORM3', 'relativism', 'I_TRUSTARMY', 'I_TRUSTPOLICE', 'I_TRUSTCOURTS',
 'scepticism', 'I_INDEP', 'I_IMAGIN', 'I_NONOBED',  'autonomy',  'I_WOMJOB', 'I_WOMPOL', 'I_WOMEDU', 'equality',
 'I_HOMOLIB', 'I_ABORTLIB', 'I_DIVORLIB', 'choice', 'I_VOICE1', 'I_VOICE2', 'I_VOI2_00', 'voice', 'WEIGHT1A',  'WEIGHT2A', 
 'WEIGHT3A',  'WEIGHT4A', 'WEIGHT1B',  'WEIGHT2B',  'WEIGHT3B', 'WEIGHT4B'], axis=1)

In [29]:
cleaning_w6.shape

(89565, 316)

In [30]:
# dropping interviewer observations and other items with high proportion of missings
cleaning_w6 = cleaning_w6.drop(['V228', # items assessing country-specific political party preferences
 'V228_LOCAL', 'V228A', 'V228B', 'V228C', 'V228D', 'V228E', 'V228F', 'V228G', 'V228H', 'V228I', 'V228J', 'V228K',

# keep these employment, income, and savings items
# 'V230', 'V231', 'V232', 'V233', 'V234', 'V236', 'V237', 'V239' (# scale of incomes)
                                
 'V241', # year of birth
                                
# these X003 items are age groups; drop them since I'll use and scale the actual age values (from V242) anyway
 'X003R2', 'X003R',
                                
# keep these immigrant status items
# 'V243', 'V244', 'V245', 'V246', 'V247'
# keeping highest education level ('V248') 
                                
 'V249',
# 'V250', # respondent lives with parents
# items assessing whether respondent was interested; would be nice/interesting to include, but missings here would be
# difficult to impute
 'V251',
 'V252',
 'V253', # size of town; interesting, but arguably hard to impute missing values
 'V254', # ethnic group
 'V255', # literacy; difficult to impute
                                
# region where interview was conducted
 'N_REGION_ISO',
 'V256', 
 'V257', # interview language
 'V260', # questionnaire version
 'V261'], # date of interview
axis=1)

In [31]:
print(cleaning_w6.shape)

(89565, 289)


In [32]:
cleaning_w6.head()

V1  V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA  V3   V4   V5   V6   V7   V8  \
0   6  12   12  615         ALG             DZA   1  1.0  1.0  1.0  NaN  1.0   
1   6  12   12  615         ALG             DZA   2  1.0  2.0  3.0  4.0  2.0   
2   6  12   12  615         ALG             DZA   3  1.0  3.0  2.0  4.0  2.0   
3   6  12   12  615         ALG             DZA   4  1.0  1.0  3.0  4.0  3.0   
4   6  12   12  615         ALG             DZA   5  1.0  1.0  1.0  2.0  1.0   

    V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  V21  V22  V23  \
0  1.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  8.0   
1  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  5.0   
2  1.0  2.0  2.0  2.0  2.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  1.0  2.0  4.0   
3  1.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0  1.0  2.0  8.0   
4  1.0  1.0  3.0  2.0  1.0  2.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  2.0  8.0   

   V24  V25  V26  V27  V28  V29  V30  V31  V32  V33  V34  V35  V36  V37  V38  \
0  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
3  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  2.0   
4  2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  2.0  1.0   

   V39  V40  V41  V42  V43  V44  V45  V46  V47  V48  V49  V50  V51  V52  V53  \
0  2.0  1.0  2.0  2.0  1.0  1.0  1.0  NaN  1.0  1.0  1.0  1.0  1.0  1.0  1.0   
1  2.0  2.0  2.0  2.0  2.0  2.0  1.0  NaN  1.0  1.0  2.0  2.0  1.0  2.0  3.0   
2  2.0  2.0  1.0  1.0  1.0  2.0  1.0  NaN  2.0  3.0  1.0  1.0  1.0  1.0  1.0   
3  2.0  2.0  2.0  2.0  1.0  2.0  2.0  NaN  1.0  1.0  1.0  2.0  2.0  2.0  2.0   
4  1.0  2.0  2.0  1.0  1.0  2.0  1.0  NaN  1.0  1.0  1.0  2.0  1.0  3.0  2.0   

   V54  V55  V56  V56_NZ  V57  V58   V59  V60  V61  V62  V63  V64  V65  V66  \
0  1.0  7.0  4.0     NaN  6.0  0.0  10.0  2.0  3.0  1.0  3.0  3.0  1.0  1.0   
1  2.0  6.0  8.0     NaN  6.0  0.0  10.0  2.0  1.0  2.0  3.0  4.0  3.0  1.0   
2  1.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  4.0  1.0  2.0  1.0  4.0  1.0   
3  2.0  6.0  8.0     NaN  6.0  0.0   6.0  2.0  1.0  3.0  1.0  4.0  3.0  1.0   
4  2.0  6.0  6.0     NaN  1.0  3.0   4.0  2.0  1.0  2.0  3.0  4.0  2.0  1.0   

   V67  V68  V69  V70  V71  V72  V73  V74  V74B  V75  V76  V77  V78  V79  V80  \
0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0   2.0  2.0  1.0  2.0  3.0  3.0  2.0   
1  1.0  1.0  2.0  2.0  3.0  2.0  1.0  1.0   1.0  3.0  2.0  2.0  2.0  2.0  2.0   
2  2.0  2.0  2.0  1.0  2.0  1.0  1.0  4.0   3.0  1.0  2.0  2.0  2.0  2.0  2.0   
3  1.0  1.0  1.0  1.0  2.0  3.0  2.0  3.0   3.0  1.0  1.0  2.0  2.0  3.0  1.0   
4  2.0  1.0  1.0  2.0  5.0  1.0  2.0  3.0   1.0  4.0  3.0  2.0  2.0  3.0  1.0   

   V81  V82  V83  V84  V85  V86  V87  V88  V89  V95  V96  V97  V98  V99  V100  \
0  1.0  2.0  2.0  4.0  NaN  NaN  NaN  NaN  NaN  5.0  8.0  7.0  6.0  8.0   7.0   
1  1.0  1.0  2.0  3.0  NaN  NaN  NaN  NaN  NaN  5.0  7.0  5.0  5.0  4.0   4.0   
2  1.0  2.0  2.0  2.0  NaN  NaN  3.0  3.0  3.0  6.0  7.0  7.0  7.0  5.0   7.0   
3  1.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  6.0  9.0  5.0  6.0  4.0   6.0   
4  2.0  2.0  2.0  2.0  NaN  NaN  NaN  NaN  NaN  5.0  8.0  4.0  7.0  4.0   6.0   

   V101  V102  V103  V104  V105  V106  V107  V108  V109  V110  V111  V112  \
0   5.0   2.0   2.0   2.0   2.0   2.0   2.0   1.0   3.0   3.0   2.0   3.0   
1   6.0   1.0   3.0   3.0   3.0   3.0   3.0   1.0   1.0   1.0   2.0   2.0   
2   5.0   1.0   3.0   3.0   4.0   4.0   4.0   3.0   2.0   2.0   2.0   3.0   
3   8.0   1.0   3.0   3.0   2.0   2.0   3.0   2.0   3.0   4.0   2.0   3.0   
4   6.0   2.0   2.0   3.0   4.0   2.0   3.0   2.0   3.0   3.0   2.0   4.0   

   V113  V114  V115  V116  V117  V118  V119  V120  V121  V122  V123  V124  \
0   2.0   2.0   4.0   3.0   3.0   2.0   3.0   2.0   3.0   4.0   4.0   3.0   
1   3.0   1.0

In [33]:
cleaning_w6.describe()

V1            V2           V2A           cow            V3  \
count  89565.0  89565.000000  89565.000000  89565.000000  8.956500e+04   
mean       6.0    471.966147    486.231943    494.832155  3.994326e+04   
std        0.0    245.114798    251.387109    251.203268  1.918431e+05   
min        6.0     12.000000     12.000000      2.000000  1.000000e+00   
25%        6.0    276.000000    332.000000    349.000000  3.910000e+02   
50%        6.0    434.000000    484.000000    560.000000  8.120000e+02   
75%        6.0    702.000000    710.000000    705.000000  1.356000e+03   
max        6.0    887.000000    901.000000    920.000000  1.394604e+06   

                 V4            V5            V6            V7            V8  \
count  89224.000000  89020.000000  88530.000000  87945.000000  88046.000000   
mean       1.094537      1.673961      1.888693      2.624129      1.516605   
std        0.346876      0.734792      0.837273      0.985910      0.796788   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        1.000000      1.000000      1.000000      2.000000      1.000000   
50%        1.000000      2.000000      2.000000      3.000000      1.000000   
75%        1.000000      2.000000      2.000000      3.000000      2.000000   
max        4.000000      4.000000      4.000000      4.000000      4.000000   

                 V9           V10           V11           V12           V13  \
count  88217.000000  88807.000000  89239.000000  89551.000000  89559.000000   
mean       1.887856      1.854291      2.094499      1.485835      1.396242   
std        1.054521      0.741204      0.851710      0.499802      0.489118   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        1.000000      1.000000      1.000000      1.000000      1.000000   
50%        1.000000      2.000000      2.000000      1.000000      1.000000   
75%        3.000000      2.000000      3.000000      2.000000      2.000000   
max        4.000000      4.000000      4.000000      2.000000      2.000000   

                V14           V15           V16           V17           V18  \
count  89558.000000  89548.000000  89556.000000  89552.000000  89553.000000   
mean       1.287121      1.770000      1.317031      1.605894      1.604871   
std        0.452421      0.420835      0.465322      0.488661      0.488881   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        1.000000      2.000000      1.000000      1.000000      1.000000   
50%        1.000000      2.000000      1.000000      2.000000      2.000000   
75%        2.000000      2.000000      2.000000      2.000000      2.000000   
max        2.000000      2.000000      2.000000      2.000000      2.000000   

                V19           V20           V21           V22           V23  \
count  89547.000000  89550.000000  89546.000000  89562.000000  88979.000000   
mean       1.590349      1.659296      1.583019      1.731281      6.834219   
std        0.491772      0.473948      0.493062      0.443296      2.274908   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        1.000000      1.000000      1.000000      1.000000      5.000000   
50%        2.000000      2.000000      2.000000      2.000000      7.000000   
75%        2.000000      2.000000      2.000000      2.000000      8.000000   
max        2.000000      2.000000      2.000000      2.000000     10.000000   

                V24           V25           V26           V27           V28  \
count  87177.000000  88725.000000  88642.000000  88572.000000  86354.000000   
mean       1.752985      0.513959      0.328806      0.237547      0.183431   
std        0.431278      0.781936      0.664859      0.574433      0.487573   
min        1.000000      0.000000      0.000000      0.000000      0.000000   
25%        2.000000      0.000000      0.000000      0.000000      0.000000   
50%        2.000000    

#### Additional variables that may need to be dropped:
- 'V258A' (sample weights for "splitted samples"; note, Germany is the only country in this data set that appears to have a splitted sample); when using weights, use 'V258' instead
- 'V2A' (use 'V2' instead) for country code
- 'C_COW_ALPHA'
- 'V144' (religious affiliation; I can't find the codes for these values, so I'm not sure how to interpret these); keep 'V144G' instead
- 'V1' (wave; there shouldn't be any variation here as these are all wave 6 data)
- 'V3' (interview number; unnecessary identifier)
- 'V23' (satisfaction with life question; redundant because using V10 as target variable instead)
- 'V262' (survey year)
- 'V247' (language spoken at home)

#### Variables to un-drop:
- Return V56_NZ to the data set (**done!**)
    - Next, use V56_NZ to fill V56 missing values if 'C_COW_ALPHA' = 'NEW'? (justify why or why not)
  
#### Thinking ahead to future steps:
- items may need to be normalized or re-scaled so that the ranges are more similar
- items may need to be reverse-coded to assist with interpretability for linear regression
- as mentioned earlier, create a new variable for V125 and V215 items
- retain and rename B_COUNTRY_ALPHA for country labels
- recode age variable 'V242'; create age categories based on groupings identified here: https://www.cia.gov/the-world-factbook/field/age-structure/

In [34]:
# what's the difference between V2 and V2A for countries that have splitted samples? Germany is the only one I'm seeing
# for this wave; there's variation in V2A values, but not in V2, COW_ALPHA or COUNTRY_ALPHA distinctions
# may not be worth it to hang on to splitted samples for this survey wave
cleaning_w6[cleaning_w6['C_COW_ALPHA'] == 'GMY'].sample(50)

V1   V2  V2A  cow C_COW_ALPHA B_COUNTRY_ALPHA    V3   V4   V5   V6  \
21713   6  276  900  255         GMY             DEU   672  1.0  1.0  1.0   
20178   6  276  900  255         GMY             DEU    63  1.0  2.0  2.0   
21886   6  276  900  255         GMY             DEU   989  3.0  1.0  1.0   
21951   6  276  901  255         GMY             DEU  1978  1.0  1.0  1.0   
20288   6  276  901  255         GMY             DEU  1154  2.0  1.0  1.0   
20146   6  276  901  255         GMY             DEU  1395  1.0  2.0  2.0   
21467   6  276  901  255         GMY             DEU  1757  1.0  1.0  1.0   
20198   6  276  900  255         GMY             DEU   173  1.0  1.0  2.0   
21449   6  276  901  255         GMY             DEU  1424  1.0  2.0  2.0   
20194   6  276  900  255         GMY             DEU   482  1.0  2.0  1.0   
20410   6  276  900  255         GMY             DEU    96  1.0  1.0  3.0   
22102   6  276  901  255         GMY             DEU  1938  1.0  2.0  2.0   
21834   6  276  901  255         GMY             DEU  1839  1.0  2.0  1.0   
20443   6  276  900  255         GMY             DEU    69  2.0  1.0  2.0   
20637   6  276  900  255         GMY             DEU    67  1.0  2.0  2.0   
21884   6  276  901  255         GMY             DEU  1992  2.0  2.0  2.0   
21497   6  276  900  255         GMY             DEU   563  1.0  1.0  2.0   
21725   6  276  901  255         GMY             DEU  1537  1.0  1.0  1.0   
21575   6  276  901  255         GMY             DEU  1475  1.0  2.0  2.0   
21562   6  276  901  255         GMY             DEU  1827  1.0  2.0  3.0   
21167   6  276  900  255         GMY             DEU   798  2.0  1.0  2.0   
21904   6  276  901  255         GMY             DEU  2002  1.0  4.0  3.0   
21122   6  276  901  255         GMY             DEU  1350  1.0  2.0  3.0   
20664   6  276  900  255         GMY             DEU    71  2.0  1.0  1.0   
20388   6  276  900  255         GMY             DEU   132  1.0  2.0  2.0   
21945   6  276  900  255         GMY             DEU   942  1.0  2.0  3.0   
21823   6  276  901  255         GMY             DEU  1627  1.0  2.0  2.0   
20933   6  276  900  255         GMY             DEU   796  1.0  1.0  1.0   
20712   6  276  901  255         GMY             DEU  1281  1.0  3.0  1.0   
20317   6  276  901  255         GMY             DEU  1124  2.0  2.0  2.0   
21235   6  276  900  255         GMY             DEU   460  1.0  2.0  1.0   
21865   6  276  901  255         GMY             DEU  1740  1.0  1.0  1.0   
20686   6  276  900  255         GMY             DEU   914  1.0  2.0  3.0   
21114   6  276  901  255         GMY             DEU  1358  2.0  2.0  3.0   
21181   6  276  901  255         GMY             DEU  1413  2.0  2.0  2.0   
20207   6  276  900  255         GMY             DEU   217  1.0  1.0  3.0   
20697   6  276  900  255         GMY             DEU   184  1.0  1.0  1.0   
22085   6  276  901  255         GMY             DEU  1931  1.0  1.0  2.0   
20340   6  276  901  255         GMY             DEU  1091  1.0  1.0  2.0   
21238   6  276  901  255         GMY             DEU  1399  1.0  1.0  1.0   
20434   6  276  900  255         GMY             DEU    87  1.0  2.0  1.0   
21353   6  276  900  255         GMY             DEU   561  1.0  1.0  2.0   
20392   6  276  900  255         GMY             DEU   224  1.0  1.0  1.0   
20217   6  276  901  255         GMY             DEU  1265  2.0  3.0  2.0   
20897   6  276  901  255         GMY             DEU  1203  1.0  2.0  2.0   
21779   6  276  900  255         GMY             DEU   747  1.0  1.0  1.0   
21716   6  276  900  255         GMY             DEU   720  1.0  1.0  1.0   
20451   6  276  900  255         GMY             DEU    23  1.0  2.0  3.0   
21299   6  276  900  255         GMY             DEU   866  1.0  1.0  1.0   
20739   6  276  901  255         GMY             DEU  1189  1.0  1.0  2.0   

        V7   V8   V9  V10  V11  V12  V13  V14  V15  V16  V17  V18  V19  V20  \


In [35]:
# how many respondents from various countries did not respond to V56?
# are these values missing at random, or en masse for some samples?
cleaning_w6[~cleaning_w6.V56.notnull()]['C_COW_ALPHA'].value_counts()

NEW    841
JPN    216
CHN    179
RUS    147
NTH    116
MOR     88
LIB     87
UZB     75
YEM     73
KUW     63
SPN     60
ALG     57
ARM     49
TAW     45
PER     42
URU     41
GRG     37
SAF     36
POL     34
EST     32
ROM     32
TUN     31
CHL     31
USA     27
ARG     26
IND     25
AUL     25
BRA     24
SWD     23
GMY     19
COL     18
TUR     16
SLV     15
MEX     13
IRQ     13
HAI      9
ECU      8
QAT      8
PSE      7
CYP      7
TRI      6
ROK      5
JOR      5
THI      4
BLR      3
SIN      2
HKG      2
LEB      2
KYR      2
Name: C_COW_ALPHA, dtype: int64

In [36]:
# how many respondents from various countries responded to V56_NZ?
wvs_w6[wvs_w6.V56_NZ.notnull()]['C_COW_ALPHA'].value_counts()

# V56 is missing entirely for repondents from New Zeland; all responses to V56_NZ are from New Zeland
# responses have a different range/scale, but could possibly be meaningfully combined into the same variable...
# justify why or why not do this

NEW    783
Name: C_COW_ALPHA, dtype: int64

In [37]:
# what to do about age?
cleaning_w6.V242.value_counts()

# create age groups based on CIA.gov categories

25.0     2383
30.0     2362
23.0     2309
20.0     2252
24.0     2215
22.0     2147
40.0     2125
26.0     2112
35.0     2092
27.0     2069
28.0     2064
18.0     2062
19.0     1999
32.0     1983
21.0     1981
29.0     1911
31.0     1886
45.0     1854
42.0     1821
36.0     1815
50.0     1805
38.0     1804
33.0     1787
34.0     1660
37.0     1633
39.0     1606
43.0     1604
41.0     1532
46.0     1524
44.0     1511
47.0     1494
55.0     1486
52.0     1463
48.0     1438
49.0     1414
60.0     1405
51.0     1373
54.0     1288
53.0     1231
56.0     1220
58.0     1213
62.0     1157
57.0     1147
65.0     1024
63.0     1008
61.0      994
59.0      967
64.0      898
70.0      839
66.0      822
68.0      723
67.0      675
72.0      647
69.0      602
73.0      597
71.0      560
74.0      511
75.0      488
76.0      432
77.0      363
78.0      297
80.0      262
79.0      257
81.0      232
82.0      173
83.0      164
84.0      153
85.0      137
17.0       56
86.0       48
87.0       42
89.0  

In [54]:
# how many missings by country?
grouped_missings = cleaning_w6.groupby('C_COW_ALPHA')
-grouped_missings.count().sub(grouped_missings.size(), 0)

V1  V2  V2A  cow  B_COUNTRY_ALPHA  V3  V4  V5   V6   V7   V8  \
C_COW_ALPHA                                                                 
ALG           0   0    0    0                0   0   5  11   18   63   31   
ARG           0   0    0    0                0   0   2   2    6   14   17   
ARM           0   0    0    0                0   0   7   5   26    7   18   
AUL           0   0    0    0                0   0   8  51   56   72   69   
AZE           0   0    0    0                0   0   0   0    0    0    0   
BLR           0   0    0    0                0   0   6  13   23   20   14   
BRA           0   0    0    0                0   0   1   3    6   11    3   
CHL           0   0    0    0                0   0   0  11    8   12   25   
CHN           0   0    0    0                0   0  29  51  101  190  107   
COL           0   0    0    0                0   0   0   2    2    7    2   
CYP           0   0    0    0                0   0   1   1    1   12    3   
ECU           0   0    0    0                0   0   0   0    1    0    0   
EGY           0   0    0    0                0   0   0   0    0    0    0   
EST           0   0    0    0                0   0   1   4    8    7   17   
GHA           0   0    0    0                0   0   0   0    0    0    0   
GMY           0   0    0    0                0   0   6   0    6    2   98   
GRG           0   0    0    0                0   0   0   0    1    1    1   
HAI           0   0    0    0                0   0  13  18   13   18   11   
HKG           0   0    0    0                0   0   2   0    5    7    2   
IND           0   0    0    0                0   0  17  67   74  166   69   
IRQ           0   0    0    0                0   0   0   2    9   11    2   
JOR           0   0    0    0                0   0   0   0    0    5   19   
JPN           0   0    0    0                0   0  39  54   69  177  132   
KUW           0   0    0    0                0   0  12  19   45   78   37   
KYR           0   0    0    0                0   0   0   0    0    0    0   
KZK           0   0    0    0                0   0   0   0    0    0    0   
LEB           0   0    0    0                0   0  55  16   22   67   34   
LIB           0   0    0    0                0   0   4  32   53   83   49   
MAL           0   0    0    0                0   0   0   0    0    1    0   
MEX           0   0    0    0                0   0   0   1    1    1    0   
MOR           0   0    0    0                0   0   2   4   81  108    9   
NEW           0   0    0    0                0   0  13  24   32   40   49   
NIG           0   0    0    0                0   0   0   0    0    1    0   
NTH           0   0    0    0                0   0  42  13   19   28   73   
PAK           0   0    0    0                0   0   1   2   16    8   19   
PER           0   0    0    0                0   0   0  13   20   32   19   
PHI           0   0    0    0                0   0   0   1    3    7    1   
POL           0   0    0    0                0   0   6   3   13    6   25   
PSE           0   0    0    0                0   0   1   2   11   15   29   
QAT           0   0    0    0                0   0   0   0    1    4    0   
ROK           0   0    0    0                0   0   1   3    6   12    9   
ROM           0   0    0    0                0   0   3   2    9   19    9   
RUS           0   0    0    0                0   0  14  21   46   62  191   
RWA           0   0    0    0                0   0   0   0    0    0    0   
SAF           0   0    0    0                0   0   5  11   32   62  125   
SIN           0   0    0    0                0   0   0   0    0    0    1   
SLV           0   0    0    0                0   0   5   1    8    4   13   
SPN           0   0    0    0                0   0   1  14   10    9   14   
SWD           0   0    0    0                0   0   3   2   14    9   67   
TAW           0   0    0    0                0   0   7  15   21   28   18   
THI      

In [60]:
# explore proportion of missings by country
# step 1: create a new dataframe
cleaning_missing_pct = cleaning_w6.isna()

In [63]:
# overwrite True/False values with underlying country codes to then groupby country
cleaning_missing_pct['C_COW_ALPHA'] = cleaning_w6['C_COW_ALPHA']

In [64]:
# check to see that things are looking the way they should
cleaning_missing_pct.head(2)

V1     V2    V2A    cow C_COW_ALPHA  B_COUNTRY_ALPHA     V3     V4  \
0  False  False  False  False         ALG            False  False  False   
1  False  False  False  False         ALG            False  False  False   

      V5     V6     V7     V8     V9    V10    V11    V12    V13    V14  \
0  False  False   True  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

     V15    V16    V17    V18    V19    V20    V21    V22    V23    V24  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

     V25    V26    V27    V28    V29    V30    V31    V32    V33    V34  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

     V35    V36    V37    V38    V39    V40    V41    V42    V43    V44  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

     V45   V46    V47    V48    V49    V50    V51    V52    V53    V54    V55  \
0  False  True  False  False  False  False  False  False  False  False  False   
1  False  True  False  False  False  False  False  False  False  False  False   

     V56  V56_NZ    V57    V58    V59    V60    V61    V62    V63    V64  \
0  False    True  False  False  False  False  False  False  False  False   
1  False    True  False  False  False  False  False  False  False  False   

     V65    V66    V67    V68    V69    V70    V71    V72    V73    V74  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

    V74B    V75    V76    V77    V78    V79    V80    V81    V82    V83  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

     V84   V85   V86   V87   V88   V89    V95    V96    V97    V98    V99  \
0  False  True  True  True  True  True  False  False  False  False  False   
1  False  True  True  True  True  True  False  False  False  False  False   

    V100   V101   V102   V103   V104   V105   V106   V107   V108   V109  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

    V110   V111   V112   V113   V114   V115   V116   V117   V118   V119  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

    V120   V121   V122   V123   V124  V125_00  V125_01  V125_02  V125_03  \
0  False  False  False  False  False     True     True     True    False   
1  False  False  False  False  False     True     True     True    False   

   V125_04  V125_05  V125_06  V125_07  V125_08  V125_09  V125_10  V125_11  \
0     True     True     True     True     True     True     True     True   
1     True     True     True     True     True     True     True     True   

   V125_12  V125_13  V125_14  V125_15  V125_16  V125_17   V126   V127   V128  \
0     True     True     True     True    False    False  False  False  False   
1     True     True     True     True    False    False  False  False  False   

    V129   V130   V131   V132   V133   V134   V135   V136   V137   V138  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

    V139   V140   V141   V142   V143   V144  V144G   V145   V146   V147  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   

    V148   V149   V150   V151   V152   V153   V154   V155   V156   V157  \
0  False  False   True   True  Fals

In [74]:
# solution: how many missings by country for each variable?
missing_grpby_country = (cleaning_missing_pct.groupby('C_COW_ALPHA').mean()*100).round(decimals=2)
missing_grpby_country

V1   V2  V2A  cow  B_COUNTRY_ALPHA   V3    V4    V5    V6    V7  \
C_COW_ALPHA                                                                     
ALG          0.0  0.0  0.0  0.0              0.0  0.0  0.42  0.92  1.50  5.25   
ARG          0.0  0.0  0.0  0.0              0.0  0.0  0.19  0.19  0.58  1.36   
ARM          0.0  0.0  0.0  0.0              0.0  0.0  0.64  0.45  2.36  0.64   
AUL          0.0  0.0  0.0  0.0              0.0  0.0  0.54  3.45  3.79  4.87   
AZE          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.00   
BLR          0.0  0.0  0.0  0.0              0.0  0.0  0.39  0.85  1.50  1.30   
BRA          0.0  0.0  0.0  0.0              0.0  0.0  0.07  0.20  0.40  0.74   
CHL          0.0  0.0  0.0  0.0              0.0  0.0  0.00  1.10  0.80  1.20   
CHN          0.0  0.0  0.0  0.0              0.0  0.0  1.26  2.22  4.39  8.26   
COL          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.13  0.13  0.46   
CYP          0.0  0.0  0.0  0.0              0.0  0.0  0.10  0.10  0.10  1.20   
ECU          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.08  0.00   
EGY          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.00   
EST          0.0  0.0  0.0  0.0              0.0  0.0  0.07  0.26  0.52  0.46   
GHA          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.00   
GMY          0.0  0.0  0.0  0.0              0.0  0.0  0.29  0.00  0.29  0.10   
GRG          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.08  0.08   
HAI          0.0  0.0  0.0  0.0              0.0  0.0  0.65  0.90  0.65  0.90   
HKG          0.0  0.0  0.0  0.0              0.0  0.0  0.20  0.00  0.50  0.70   
IND          0.0  0.0  0.0  0.0              0.0  0.0  0.42  1.64  1.81  4.07   
IRQ          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.17  0.75  0.92   
JOR          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.42   
JPN          0.0  0.0  0.0  0.0              0.0  0.0  1.60  2.21  2.82  7.25   
KUW          0.0  0.0  0.0  0.0              0.0  0.0  0.92  1.46  3.45  5.99   
KYR          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.00   
KZK          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.00   
LEB          0.0  0.0  0.0  0.0              0.0  0.0  4.58  1.33  1.83  5.58   
LIB          0.0  0.0  0.0  0.0              0.0  0.0  0.19  1.50  2.49  3.89   
MAL          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.08   
MEX          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.05  0.05  0.05   
MOR          0.0  0.0  0.0  0.0              0.0  0.0  0.17  0.33  6.75  9.00   
NEW          0.0  0.0  0.0  0.0              0.0  0.0  1.55  2.85  3.80  4.76   
NIG          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.06   
NTH          0.0  0.0  0.0  0.0              0.0  0.0  2.21  0.68  1.00  1.47   
PAK          0.0  0.0  0.0  0.0              0.0  0.0  0.08  0.17  1.33  0.67   
PER          0.0  0.0  0.0  0.0              0.0  0.0  0.00  1.07  1.65  2.64   
PHI          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.08  0.25  0.58   
POL          0.0  0.0  0.0  0.0              0.0  0.0  0.62  0.31  1.35  0.62   
PSE          0.0  0.0  0.0  0.0              0.0  0.0  0.10  0.20  1.10  1.50   
QAT          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.09  0.38   
ROK          0.0  0.0  0.0  0.0              0.0  0.0  0.08  0.25  0.50  1.00   
ROM          0.0  0.0  0.0  0.0              0.0  0.0  0.20  0.13  0.60  1.26   
RUS          0.0  0.0  0.0  0.0              0.0  0.0  0.56  0.84  1.84  2.48   
RWA          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.00   
SAF          0.0  0.0  0.0  0.0              0.0  0.0  0.14  0.31  0.91  1.76   
SIN          0.0  0.0  0.0  0.0              0.0  0.0  0.00  0.00  0.00  0.00   
SLV          0.0  0.0  0.0  0.0              0.0  0.0  0.47  0.09  0.75  0.37   
SPN          0.0  0.0  0.0  0.0              

In [72]:
# what's the largest proportion of missing values for any given country for each variable?
((cleaning_missing_pct.groupby('C_COW_ALPHA').mean()*100).max()).round(decimals=2)

V1                   0.00
V2                   0.00
V2A                  0.00
cow                  0.00
B_COUNTRY_ALPHA      0.00
V3                   0.00
V4                   4.58
V5                   3.45
V6                   6.75
V7                   9.00
V8                   7.64
V9                  13.55
V10                  5.50
V11                  2.18
V12                  0.44
V13                  0.12
V14                  0.16
V15                  0.56
V16                  0.24
V17                  0.40
V18                  0.36
V19                  0.60
V20                  0.48
V21                  0.64
V22                  0.10
V23                  2.84
V24                 10.80
V25                 16.04
V26                 17.11
V27                 17.65
V28                100.00
V29                100.00
V30                 19.49
V31                100.00
V32                 16.04
V33                 19.57
V34                 22.41
V35                100.00
V36         

Notes:
- features to drop (variables where 100% of values are missing for any given sample (cannot impute) or too many missings or simply unnecessary):
    - V28, V29, V31: membership items
    - V35: organizational membership
    - V36-V43: attitudes towards neighbors 
    - V46: attitudes when jobs are scarce 
    - V50: opinion on how woman working impacts children 
    - V85-V89: willingness to take political action 
    - V95: political leanings
    - V109: confidence in armed forces; 100% missing for Egypt only (N = 1523)
    - V112: confidence in labor unions
    - V118: confidence in civil service
    - V125_01-V125_17: confidence in political/economic union variables
    - V127-V130: political system variables
    - V135: essential characteristic of democracy; 100% missing for Egypt only (N = 1523)
    - V141: how democratic is govt; missing for multiple countries
    - V142: respect for human rights; missing for Uzbekistan only (N = 1500); would be nice to keep, but arguably not crucial
    - V145-V146, then V148-149: religious beliefs and practices; KEEP V147 (how religious are you)
    
- cases to drop (participating countries) due to too many missings:
    - Egypt (N = 1523)

- feature engineering notes: 
    - V56 and V56_NZ: people would try to take advantage; consolidate into one variable
    - V58: number of children; missing entirely for Hong Kong...contemplating just dropping them (N = 1000) to preserve this feature; alternatively I could pull in birthrate and other health/wellbeing data for all countries...we'll see; TBD
    - V60-V65: prioritizing country aims; missing entirely for Kuwait...contemplating dropping them as well (N = 1303)
    - V74 and V74B: Schwartz helpful value; consolidate into one variable based on whichever has fewer missings???
    - V81: priorities (protecting environment vs economic growth); 100% missing for TRI (N = 999) only; considering dropping them
    - V102-107: trust in various groups (family, neighborhood, etc.) items. New Zealand is the only country that has missings across the board (N = 841); contemplating dropping them
    

In [89]:
# what's going on with V74 and V74B for respondents from GMY? 
# it looks like they're responding to one item OR the other...maybe I can combine these into the same variable
cleaning_w6[cleaning_w6['C_COW_ALPHA'] == 'GMY'][['V74', 'V74B']].sample(20)

V74  V74B
21034  1.0   NaN
21492  3.0   NaN
22127  NaN   4.0
21429  NaN   3.0
21534  4.0   NaN
22051  NaN   1.0
20215  4.0   NaN
20918  3.0   NaN
21840  5.0   NaN
21858  4.0   NaN
22018  4.0   NaN
20632  NaN   2.0
20134  5.0   NaN
20138  NaN   2.0
20282  NaN   NaN
21893  2.0   NaN
20595  3.0   NaN
20570  NaN   1.0
22138  2.0   NaN
21341  2.0   NaN

In [87]:
# how do V74 and V74B look for respondents from other countries with 100% missings on one or the other?
# countries where this is the case: from top - ARG, ARM, AUL, AZE, BLR, CHL...from bottom - ZIM, UZB, USA, URU, UKR, TUR
# it looks like they're responding to one item OR the other...maybe I can combine these into the same variable
cleaning_w6[cleaning_w6['C_COW_ALPHA'] == 'ZIM'][['V74', 'V74B']].sample(20)

V74  V74B
88396  2.0   NaN
89396  1.0   NaN
88417  1.0   NaN
88252  1.0   NaN
88074  2.0   NaN
88531  2.0   NaN
88926  1.0   NaN
89075  1.0   NaN
88824  1.0   NaN
89115  2.0   NaN
88341  2.0   NaN
88132  1.0   NaN
89276  3.0   NaN
89061  3.0   NaN
88655  2.0   NaN
88085  1.0   NaN
88748  2.0   NaN
89087  4.0   NaN
88452  2.0   NaN
88722  1.0   NaN

In [73]:
# sample sizes for reference
cleaning_w6.groupby('C_COW_ALPHA')['V3'].count()

C_COW_ALPHA
ALG    1200
ARG    1030
ARM    1100
AUL    1477
AZE    1002
BLR    1535
BRA    1486
CHL    1000
CHN    2300
COL    1512
CYP    1000
ECU    1202
EGY    1523
EST    1533
GHA    1552
GMY    2046
GRG    1202
HAI    1996
HKG    1000
IND    4078
IRQ    1200
JOR    1200
JPN    2443
KUW    1303
KYR    1500
KZK    1500
LEB    1200
LIB    2131
MAL    1300
MEX    2000
MOR    1200
NEW     841
NIG    1759
NTH    1902
PAK    1200
PER    1210
PHI    1200
POL     966
PSE    1000
QAT    1060
ROK    1200
ROM    1503
RUS    2500
RWA    1527
SAF    3531
SIN    1972
SLV    1069
SPN    1189
SWD    1206
TAW    1238
THI    1200
TRI     999
TUN    1205
TUR    1605
UKR    1500
URU    1000
USA    2232
UZB    1500
YEM    1000
ZIM    1500
Name: V3, dtype: int64